src/patterns.jl

```diff
@@ -0,0 +1,112 @@
+"""
+    Pattern
+
+A pattern is just a wrapper around a `Tuple{Vararg{Symbol}}` which enables searching and
+filtering for matching components and dimension paths in a [`Dataset`](@ref).
+Special symbols `:_` and `:__` are used as wildcards, similar to `*` and `**` in glob
+pattern matching.
+
+# Example
+```jldoctest
+julia> using AxisSets: Pattern;
+
+julia> items = [
+           (:train, :input, :load, :time),
+           (:train, :input, :load, :id),
+           (:train, :input, :temperature, :time),
+           (:train, :input, :temperature, :id),
+           (:train, :output, :load, :time),
+           (:train, :output, :load, :id),
+       ];
+
+julia> filter(in(Pattern(:__, :time)), items)
+3-element Array{NTuple{4,Symbol},1}:
+ (:train, :input, :load, :time)
+ (:train, :input, :temperature, :time)
+ (:train, :output, :load, :time)
+
+julia> filter(in(Pattern(:__, :load, :_)), items)
+4-element Array{NTuple{4,Symbol},1}:
+ (:train, :input, :load, :time)
+ (:train, :input, :load, :id)
+ (:train, :output, :load, :time)
+ (:train, :output, :load, :id)
+```
+"""
+struct Pattern
+    segments::Tuple{Vararg{Symbol}}
+
+    # We reduce the segments on construction to remove extra wildcards
+    # (e.g., `(:_, :__, ...)`, `(:__, :_, ...`)
+    # NOTE: Since these patterns shouldn't be very long it seemed alright to do in an
+    # inner constructor
+    function Pattern(segments::Tuple{Vararg{Symbol}})
+        n = length(segments)
+        mask = trues(n)
+
+        # Prev index used for lookup up segment and mask values as we iterate through
+        j = 1
+
+        # Iterating forwards and backwards seemed like the easiest way to
+        # remove extra wildcards on either side of a `:__`
+        for i in Iterators.flatten([2:n, n-1:-1:1])
+            val = segments[i]
+            prev = segments[j]
+            prev_mask = mask[j]
+
+            if val in (:_, :__) && prev === :__ && prev_mask
+                mask[i] = false
+            else
+                j = i
+            end
+        end
+
+        return new(segments[mask])
+    end
+end
+
+Pattern(segments::Symbol...) = Pattern(segments)
+Base.convert(::Type{Pattern}, x::Tuple{Vararg{Symbol}}) = Pattern(x)
+
+function Base.in(item::Tuple{Vararg{Symbol}}, pattern::Pattern)
+    # If our pattern has more segments than the item then it isn't going to match
+    length(pattern.segments) <= length(item) || return false
+
+    # Setup our item and pattern iterators
+    item_iter = iterate(item)
+    pat_iter = iterate(pattern.segments)
+
+    # Loop as long as neither iterator is exhausted
+    while item_iter !== nothing && pat_iter !== nothing
```

Comment:
I feel like we could just convert the `pattern`/`item` and do regex matching here.

Reply:
We could, and I did consider it for a first pass, but:

1. We don't want to support the full regex syntax in this context
2. I think we want to keep things as tuple as much as possible to avoid potential character collisions or needing to escape delimiters.
3. Constructing a regex on every call to `in` is expensive. Typically, you'd compile it once and reuse it.

Reply:
FWIW, I was curious how expensive generating the regex on the fly would be.

```
julia> @benchmark regexin($item, $pattern)
BenchmarkTools.Trial:
  memory estimate:  1.73 KiB
  allocs estimate:  35
  --------------
  minimum time:     5.944 μs (0.00% GC)
  median time:      6.826 μs (0.00% GC)
  mean time:        7.659 μs (1.39% GC)
  maximum time:     3.223 ms (33.10% GC)
  --------------
  samples:          10000
  evals/sample:     5

julia> @benchmark $item in $pattern
BenchmarkTools.Trial:
  memory estimate:  128 bytes
  allocs estimate:  4
  --------------
  minimum time:     51.301 ns (0.00% GC)
  median time:      52.898 ns (0.00% GC)
  mean time:        59.882 ns (5.72% GC)
  maximum time:     1.828 μs (96.98% GC)
  --------------
  samples:          10000
  evals/sample:     982
```